source/target/target_family.h

```diff
@@ -0,0 +1,69 @@
+/**
+ * @file    target_family.h
+ * @brief   
+ *
+ * DAPLink Interface Firmware
+ * Copyright (c) 2009-2018, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TARGET_FAMILY_H
+#define TARGET_FAMILY_H
+
+#include "stdint.h"
+#include "swd_host.h"
+
+typedef enum _reset_type { 
+    kHardwareReset=1, 
+    kSoftwareReset, 
+} reset_type_t; 
+ 
+typedef struct target_family_descriptor { 
+    uint32_t family_id; 
+    reset_type_t default_reset_type; 
+    uint32_t soft_reset_type; 
+    void (*target_before_init_debug)(void); 
+    void (*prerun_target_config)(void); 
+    uint8_t (*target_unlock_sequence)(void); 
+    uint8_t (*security_bits_set)(uint32_t addr, uint8_t *data, uint32_t size); 
+    uint8_t (*target_set_state)(TARGET_RESET_STATE state); 
+    void (*swd_set_target_reset)(uint8_t asserted);
+} target_family_descriptor_t;
+
+
+#define STUB_HW_RESET_FAMILY_ID         0x01
+#define STUB_SW_VECTRESET_FAMILY_ID     0x02
+#define STUB_SW_SYSRESETREQ_FAMILY_ID   0x03
+#define NXP_KINETIS_K_SERIES_FAMILY_ID  0x11
```

Comment:
That's a really good idea to group the family IDs by vendor! Let's take that a bit further and define the top 8 bits of family ID as vendor and lower 8 bits as family (limiting family ID to 16-bit). Then add a vendor ID enum.

```c
enum _vendor_ids {
    kStub_VendorID = 0x0000,
    kNXP_VendorID = 0x0b00,
    kNordic_VendorID = 0x3600,
    kRealtek_VendorID = 0x7c00,
    // ...
};

enum _family_ids {
    kNXP_KinetisK_FamilyID = kNXP_VendorID | 0x00,
};
```

[CMSIS already defines vendor IDs](http://arm-software.github.io/CMSIS_5/Pack/html/pdsc_family_pg.html#DeviceVendorEnum), so we should use the same numbers. There's nothing above 255 yet…